core/network-libp2p/src/custom_proto/behaviour.rs

```diff
@@ -596,74 +578,53 @@ where
 		event: <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
 	) {
 		match event {
-			CustomProtosHandlerOut::CustomProtocolClosed { protocol_id, result } => {
-				let pos = self.open_protocols.iter().position(|(s, p)|
-					s == &source && p == &protocol_id
-				);
-
-				if let Some(pos) = pos {
-					self.open_protocols.remove(pos);
-				} else {
-					debug_assert!(false, "Couldn't find protocol in open_protocols");
-				}
+			CustomProtoHandlerOut::CustomProtocolClosed { result } => {
+				self.opened_peers.remove(&source);
 
-				let event = CustomProtosOut::CustomProtocolClosed {
-					protocol_id,
+				let event = CustomProtoOut::CustomProtocolClosed {
 					result,
 					peer_id: source,
 				};
 
 				self.events.push(NetworkBehaviourAction::GenerateEvent(event));
 			}
-			CustomProtosHandlerOut::CustomProtocolOpen { protocol_id, version } => {
-				debug_assert!(!self.open_protocols.iter().any(|(s, p)|
-					s == &source && p == &protocol_id
-				));
-				self.open_protocols.push((source.clone(), protocol_id));
+			CustomProtoHandlerOut::CustomProtocolOpen { version } => {
+				self.opened_peers.insert(source.clone());
 
 				let endpoint = self.connected_peers.get(&source)
 					.expect("We only receive events from connected nodes; QED").clone();
-				let event = CustomProtosOut::CustomProtocolOpen {
-					protocol_id,
+				let event = CustomProtoOut::CustomProtocolOpen {
 					version,
 					peer_id: source,
 					endpoint,
 				};
 
 				self.events.push(NetworkBehaviourAction::GenerateEvent(event));
 			}
-			CustomProtosHandlerOut::CustomMessage { protocol_id, message } => {
-				debug_assert!(self.open_protocols.iter().any(|(s, p)|
-					s == &source && p == &protocol_id
-				));
-				let event = CustomProtosOut::CustomMessage {
+			CustomProtoHandlerOut::CustomMessage { message } => {
+				let event = CustomProtoOut::CustomMessage {
```

Comment:
You could keep the debug assert as `self.is_open(&source)`

Reply:
There are two more removed asserts nearby - in `CustomProtoHandlerOut::CustomProtocolOpen` (opposite assert) && in `CustomProtoHandlerOut::Clogged` (the same assert). Probably worth keep these as well?